test/blas1_test.hpp

```diff
@@ -0,0 +1,175 @@
+#ifndef BLAS1_TEST_HPP_DFQO1OHP
+#define BLAS1_TEST_HPP_DFQO1OHP
+
+#include <cmath>
+#include <complex>
+#include <cstdlib>
+#include <iomanip>
+#include <iostream>
+#include <vector>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <interface/blas1_interface_sycl.hpp>
+
+using namespace cl::sycl;
+using namespace blas;
+
+template <typename ClassName>
+struct option_size;
+#define RANDOM_SIZE UINT_MAX
+#define RANDOM_STRD UINT_MAX
+#define REGISTER_SIZE(size, test_name)          \
+  template <>                                   \
+  struct option_size<class test_name> {         \
+    static constexpr const size_t value = size; \
+  };
+template <typename ClassName>
+struct option_strd;
+#define REGISTER_STRD(strd, test_name)          \
+  template <>                                   \
+  struct option_strd<class test_name> {         \
+    static constexpr const size_t value = strd; \
+  };
+template <typename ScalarT, typename ClassName>
+struct option_prec;
+#define REGISTER_PREC(type, val, test_name)   \
+  template <>                                 \
+  struct option_prec<type, class test_name> { \
+    static constexpr const type value = val;  \
+  };
+
+// Wraps the above arguments into one template parameter.
+// We will treat template-specialized blas_templ_struct as a single class
+template <class ScalarT_, class ExecutorType_>
+struct blas_templ_struct {
+  using scalar_t = ScalarT_;
+  using executor_t = ExecutorType_;
+};
+// A "using" shortcut for the struct
+template <class ScalarT_, class ExecutorType_ = SYCL>
+using blas1_test_args = blas_templ_struct<ScalarT_, ExecutorType_>;
+
+// the test class itself
+template <class B>
+class BLAS1_Test;
+
+template <class ScalarT_, class ExecutorType_>
+class BLAS1_Test<blas1_test_args<ScalarT_, ExecutorType_>>
+    : public ::testing::Test {
+ public:
+  using ScalarT = ScalarT_;
+  using ExecutorType = ExecutorType_;
+
+  BLAS1_Test() {}
+
+  virtual ~BLAS1_Test() {}
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+  static size_t rand_size() {
+    size_t ret = rand() >> 5;
+    int type_size =
+        sizeof(ScalarT) * CHAR_BIT - std::numeric_limits<ScalarT>::digits10 - 2;
+    return (ret & (std::numeric_limits<size_t>::max() +
+                   (size_t(1) << (type_size - 2)))) +
+           1;
+  }
+
+  template <typename DataType,
+            typename ValueType = typename DataType::value_type>
+  static void set_rand(DataType &vec, size_t _N) {
+    ValueType left(-1), right(1);
+    for (size_t i = 0; i < _N; ++i) {
+      vec[i] = ValueType(rand() % int(right - left) * 1000) * .001 - right;
```

Comment:
rand() is a temporary solution, we may want to use some other generator in future, which is to be decided.